File: src/changeResolution.py
import cv2
import time


def resize_and_grayscale(image):
    return cv2.cvtColor(cv2.resize(image, (480, 360)), cv2.COLOR_BGR2GRAY)


def set_cap_properties(vcap, width=640, height=480, fps=30):
    vcap.set(cv2.CAP_PROP_FRAME_WIDTH, width)
    vcap.set(cv2.CAP_PROP_FRAME_HEIGHT, height)
    vcap.set(cv2.CAP_PROP_FPS, fps)


cap = cv2.VideoCapture(0)
cap.set(cv2.CAP_PROP_BUFFERSIZE, 5)
set_cap_properties(cap, 1280, 720)
success, image = cap.read()

lasttime = time.monotonic().__trunc__()
while success:
    success, image = cap.read()

    cv2.imshow('Lo-Res grayscale', resize_and_grayscale(image))
    # Update high-res every second
    # if lasttime < (time.monotonic().__trunc__()):
    #     cv2.imshow('Lo-Speed Hi-res', image)
    #     lasttime = time.monotonic().__trunc__()

    if cv2.waitKey(1) & 0xFF == ord('q'):
        break

cap.release()
cv2.destroyAllWindows()

File: src/changeResolution.cpp
#include <opencv2/opencv.hpp>
#include <iostream>

using namespace cv;

VideoCapture videoCapture(0);
//VideoCapture videoCapture("/dev/video0", CAP_V4L2);

void exit_node(bool hasFaulted = false) {
    videoCapture.release();
    destroyAllWindows();
    exit(hasFaulted);
}

void crit_error(const std::string &errMsg) {
    std::cerr << errMsg << std::endl;
    exit_node(true);
}

void set_cap_properties(int width = 1280, int height = 720, int fps = 120) {
    //videoCapture.set(CAP_PROP_FOURCC ,('M', 'J', 'P', 'G') );
    videoCapture.set(CAP_PROP_FRAME_WIDTH, width);
    videoCapture.set(CAP_PROP_FRAME_HEIGHT, height);
    videoCapture.set(CAP_PROP_FPS, fps);
}

int main() {
    set_cap_properties(1920, 1080, 25);
    if (!videoCapture.isOpened()) crit_error("Error opening video stream or file");

    // Read the frames to the last frame
    while (videoCapture.isOpened()) {
        Mat frame;
        if (!videoCapture.read(frame)) crit_error("No frames in buffer, camera may have disconnected");
        Mat rsizd_frame;
        resize(frame, rsizd_frame, Size(480, 360), INTER_LINEAR);
        Mat bwFrame;
        cvtColor(rsizd_frame, bwFrame, COLOR_BGR2GRAY);
        imshow("Frame", frame);

        if (waitKey(1) == 'q') {
            exit_node();
        }
    }
    crit_error("camera connection closed unexpectedly");
}